File: python/Q76_minimum_window_substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#


# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        t_dict: dict[str, int] = dict()
        result: str | None = None
        for c in t:
            t_dict[c] = t_dict.get(c, 0) + 1

        left: int = 0
        remained: int = len(t_dict)
        counter: dict[str, int] = dict()

        for right in range(len(s)):
            letter: str = s[right]
            counter[letter] = counter.get(letter, 0) + 1

            if letter not in t_dict:
                continue

            if counter[letter] == t_dict[letter]:
                remained -= 1

            if remained == 0:
                while remained == 0:
                    l_letter: str = s[left]
                    if l_letter in t_dict and counter[l_letter] == t_dict[l_letter]:
                        remained += 1

                    counter[l_letter] -= 1
                    left += 1

                if result is None or len(result) > right - left + 2:
                    result = s[left - 1 : right + 1]

        return result if result is not None else ""


# @lc code=end

File: rust/src/leetcode/Q76_minimum_window_substring.rs
/*
 * @lc app=leetcode id=76 lang=rust
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let t_vec: Vec<char> = t.chars().collect();
        let mut t_map: HashMap<char, i32> = HashMap::new();
        for c in t_vec {
            *t_map.entry(c).or_insert(0) += 1;
        }

        let s_vec: Vec<char> = s.chars().collect();
        let mut s_map: HashMap<char, i32> = HashMap::new();

        let mut left: usize = 0;
        let mut remained: usize = t_map.len();

        let mut result_start: usize = 0;
        let mut result_length: usize = s.len() + 1;

        for right in 0..s_vec.len() {
            let right_char: char = s_vec[right];
            *s_map.entry(right_char).or_insert(0) += 1;

            if s_map.get(&right_char) == t_map.get(&right_char) {
                remained -= 1;

                while remained == 0 {
                    if right - left + 1 < result_length {
                        result_start = left;
                        result_length = right - left + 1;
                    }

                    let left_char = s_vec[left];

                    if s_map.get(&left_char) == t_map.get(&left_char) {
                        remained += 1;
                    }

                    *s_map.get_mut(&left_char).unwrap() -= 1;
                    left += 1;
                }
            }
        }

        if result_length == s.len() + 1 {
            String::new()
        } else {
            s_vec[result_start..result_start + result_length]
                .iter()
                .collect()
        }
    }
}
// @lc code=end
